[Fix-17534][Service/Master] Add global parameters and varpool from current workflow instance and add them to start params list of the trigger request of a sub workflow

kvermeulen opened the pull request:
<!--Thanks very much for contributing to Apache DolphinScheduler, we are happy that you want to help us improve DolphinScheduler! -->

## Purpose of the pull request

Close [#17534](https://github.com/apache/dolphinscheduler/issues/17534)

## Brief change log

When triggering a sub workflow, lookup all global parameters from the current, and any parent workflow instances and pass them along in the start param list of the trigger request of the sub workflow

## Verify this pull request

This change added tests and can be verified as follows:
- *Added a new integration test for validating the fixed behaviour*
- *Manually verified the change by using the test-case mentioned in issue [#17534](https://github.com/apache/dolphinscheduler/issues/17534)*

## Pull Request Notice
[Pull Request Notice](https://github.com/apache/dolphinscheduler/blob/dev/docs/docs/en/contribute/join/pull-request.md)

This is my very first pull-request, so bear with me on this one. :-)

A colleague of mine opened the original issue. This PR fixes these erroneous workflows for us, but I'm not entirely sure if this is the correct approach taken. I'll try and follow-up to make sure this PR meets all coding conventions and fulfills all requirements. I'll do my best on taking the time on improving the implementation on any pointers/advice given, if the implementation in the PR is not entirely correct or does not cover all use-cases, but I can't make any promises there.

Thank you for reviewing.

ruanwenjun commented:
3.3.1 has been closed, please submit PR to dev.

Commit:
[Fix-17534][Service/Master] also merge and pass global parameters and varpool from current workflow instance, together with the workflow instances command params to the sub workflow

kvermeulen commented:
I tried rebasing this one on dev. Is this sufficient, or should I create an entire new PR?

ruanwenjun commented:
> I tried rebasing this one on dev. Is this sufficient, or should I create an entire new PR?

No need to create a new PR.

kvermeulen commented:
Fixed the 3 issues detected by sonarqube

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/executor/plugin/subworkflow/SubWorkflowLogicTask.java:
```diff
@@ -234,7 +249,7 @@ private SubWorkflowLogicTaskRuntimeContext triggerNewSubWorkflow() {
                 .environmentCode(workflowInstance.getEnvironmentCode())
                 // todo: transport varpool and local params
-                .startParamList(commandParam.getCommandParams())
+                .startParamList(mergeParams(asList(commandParam.getCommandParams(), globalParamsOfParents)))
```
We should merge the current workflow instance's command params, global params, varpool as the subworkflow instance's command param is enough, we don't need to find the parant global params here, since the parent global params should already exist in command params.

kvermeulen replied:
Sorry, I missed that indeed. Parent global params already exist in the command params. I'll update and remove the unnecessary code. Can you maybe verify if the order of merging the params is correct? Before I didn't include the varpool for instance.

I'll update my commit and force push on the PR. I don't know if this is best practice or if we'll loose trace of the previous change. However, I do like a code changes for a fix in a single commit. Makes it more concise maybe. But any advice here is also welcome.

ruanwenjun replied:
When the code change submits a commit is better, the commit will be polished when PR merged by committer.

kvermeulen commented:
Updated and corrected the title of the PR after receiving some review remarks from @ruanwenjun, who pointed out that only the global parameters and varpool of the current workflow instance, need to be passed  when triggering a subflow. Global parameters from any parent workflow(s) will already be in the command params from the current workflow instance. (thanks for pointing out!)

kvermeulen commented:
Just to be sure, is there still any action needed from my side currently?

SbloodyS committed:
Merge branch 'dev' into Fix-17534

ruanwenjun commented on dolphinscheduler-master/src/main/java/org/apache/dolphinscheduler/server/master/engine/executor/plugin/subworkflow/SubWorkflowLogicTask.java:
```diff
@@ -222,6 +234,11 @@ private SubWorkflowLogicTaskRuntimeContext triggerNewSubWorkflow() {
+        final List<Property> paramList = mergeParams(asList(
+                commandParam.getCommandParams(),
+                new ArrayList<>(deserializeVarPool(workflowInstance.getVarPool())),
+                new ArrayList<>(deserializeVarPool(workflowInstance.getGlobalParams()))));
```
The varpool's priority should be highest.
```suggestion
                new ArrayList<>(deserializeVarPool(workflowInstance.getGlobalParams()))),
                commandParam.getCommandParams(),
                new ArrayList<>(deserializeVarPool(workflowInstance.getVarPool())));
```

kvermeulen replied:
Force pushed again because I committed a compile-time error. The varpool should have the highest priority now when merging parameters.

kvermeulen committed:
[Fix-17534][Service/Master] varpool should have highest priority

Commit:
[Fix-17534][Service/Master] fix spotless warnings

ruanwenjun reviewed (approved):
LGTM

SbloodyS reviewed (approved):
+1
